### PR TITLE
Make links in messages clickable [1/2]

### PR DESCRIPTION
**Pull request by djfpaagman**
The other day I got a message with a link in it, which I had to copy paste into my browser. I think it makes sense to make links clickable, also called "autolinking". For this I've added the [rinku](https://github.com/vmg/rinku) gem which looks a bit stale at first glance but I think it still does the job alright. Better than messing with regexes at least.

I've decided to just pass the string through `rinku` and then through `simple_format` with the rest of the arguments.

See screenshots below for what it looks like. Note that links are not really standing out right now, is that something to improve as well? I noticed that none of the links have underlines or anything so feedback there is welcome.

image
image

### Pull request checklist

- [x] Your code contains tests relevant for code you modified
- [x] You have linted and tested the project with `bin/check`

### CODE REVIEW

**Comment by joemasilotti** [comment]
I love the idea of linking to URLs and email addresses. But I'm not 100% sure we want to add a dependency that hasn't been updated in a few years. Also, if it's written in C is there a native binary that needs to be installed?

I'm wondering if we want to explore https://github.com/dejan/auto_html instead. It was last updated 2 months ago and seems to do what we want. 

Dependencies aside, in the long term messages should probably be formatted by the writer via ActionText.

**Comment by djfpaagman** [comment]
@joemasilotti I've replaced `rinku` with `auto_html` in the latest commit :+1:, it does look like a better gem indeed, I didn't know of it's existence yet, so it's nice to discover it this way as well :smile:

File: Gemfile [hunk 1]
```diff
@@ -52,3 +52,4 @@ gem "sidekiq", "~> 6.3"
 gem "sitemap_generator", "~> 6.1"
 gem "stripe", ">= 2.8", "< 6.0"
 gem "tzinfo-data", platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+gem "auto_html"
```
Comment (joemasilotti) [inline Gemfile:55]: Please lock this to a minor version via "~>" and sort this "group" of gems.

File: app/helpers/message_helper.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module MessageHelper
+  def smart_format(string, **options)
+    link_filter = AutoHtml::Link.new
+    with_links = link_filter.call(string)
+
+    simple_format(with_links, options)
```
Comment (joemasilotti) [inline app/helpers/message_helper.rb:6]: auto_html has it's own `AutoHtml::SimpleFormat` pipeline. Let's use that instead via a `AutoHtml::Pipeline`.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
Looking good! I made two small requests.

Outside of those, the README recommends [storing the formatted text in the database](https://github.com/dejan/auto_html#using-autohtml-with-activerecord). I think we should handle that now since messages are not yet paginated and some conversations have 10+ messages. This also gets us around needing a helper because we can call `message.body_html` directly.

That said, it will require a small data migration to update existing records. A rake task that I can run in production will be perfect.

**Comment by djfpaagman** [comment]
@joemasilotti implemented your suggestions, I've added a migration for the new field and a rake task to fill it. To deploy this now requires the following steps:

1. **Merge the PR** :wink: 
1. **Run the migration**
2. **Run the Rake task** `rails messages:migrate_body_html`

Note that it does not use the new field yet, because it will be empty until the migration has ran. To have a nice deployment I've decided to split it out. So that means the following needs to happen in a new PR/deploy:

1. **Start using the new field.** A small thing still to tackle is the use of the `mt-4` class for paragraphs now (see code snippet below). SimpleFormat does not support that. A possible solution could be to wrap the message in a `prose` class from tailwind typography, which handles margins for paragraphs automatically. What do you think?
https://github.com/joemasilotti/railsdevs.com/blob/61105c0363596d885ccd6b2ffa18d10b359c88ec/app/views/messages/_message.html.erb#L14
2. **Remove the `skip` in `test/integration/messages_test.rb:101`.** Since the formatting is now also tested in the model this could also be wrapped up in the "messages are formatted" test I guess.
2. (optional) Add `null: false` to the `body_html` column
1. (optional) Remove the Rake task

File: Gemfile.lock [hunk 3]
```diff
@@ -305,6 +310,7 @@ GEM
       mime-types (>= 1.16, < 4.0)
       netrc (~> 0.8)
     rexml (3.2.5)
+    rinku (2.0.6)
```
Comment (joemasilotti) [inline Gemfile.lock:313]: Does auto_html depend on rinku or did this sneak in from your first pass?

**Review by joemasilotti — APPROVED** [review]
Outside of one minor comment this looks great!

Can you also submit the follow up PR that takes care of removing the rake task, making sure the column is non-null, and uses the new column? Once that is approved I will kick off the flow.

Thanks for the contribution!

File: Gemfile.lock (hunk 3 above)
Comment (djfpaagman) [reply Gemfile.lock:313]: heh yeah interestingly it's also a dependency from `auto_html` :smile:

**Comment by djfpaagman** [comment]
Part 2 is up! #262 .

I left a comment there about the migration, I think it would be nice to tackle that before merging either of the PRs.

**Comment by joemasilotti** [comment]
OK, all merged in! Thanks so much for the contribution.

`auto_html` did take down the site for a few minutes, but I [opened a PR](https://github.com/dejan/auto_html/pull/187) so it doesn't happen to anyone else in the future.

image

**Comment by djfpaagman** [comment]
Oof weird that that didn't show up in development 🤔  great that it's merged and live though 🎉

**Comment by joemasilotti** [comment]
I think because Standard depends on it so we already had the gem installed.

**Pull request merged**